DOC Add Bernoulli, Categorical Distribution to Generalized Linear Models (GLM), add note to `Logistic Regression` on being a GLM

awinml opened the pull request:
#### Reference Issues/PRs
Fixes #24611 

#### What does this implement/fix? Explain your changes.
- Updates the description of Logistic Regression in the user guide to highlight the fact that it a regression model implemented as a classifier in scikit-learn.
- Added a note explaining how the Logistic Regression model is a special case of the Generalized Linear Regression Model (GLM).
 - Renamed `Generalized Linear Regression` Section to `Generalized Linear Models`.
 - Corrected Formula for Unit Deviance for Gamma Distribution 
 - Added section on Bernoulli and Categorical Distributions:
    - Added Unit Deviance for Bernoulli and Categorical Distributions
    - Added graph of PMF of Binomial Distribution
    - Added choice of Distribution based on Input
    - Added examples of use cases
    - Highlighted Bernoulli Distribution being a special case of Categorical Distribution

awinml committed:
Update doc/modules/linear_model.rst

awinml committed:
Merge branch 'main' into doc_logistic_regression

ArturoAmorQ commented on doc/modules/linear_model.rst:
```diff
@@ -848,25 +848,36 @@ Ridge Regression`_, see the example below.
+.. note:: **Logistic Regression as a special case of the Generalized Linear Regression Model (GLM)**
+
+    Logistic regression is a special case of
+    :ref:`generalized_linear_regression` with a Binomial/Bernoulli conditional
```
I would suggest using this opportunity to change the name of the section from "Generalized Linear Regression" to "Generalized Linear Models" in the linear_model.rst.

Maybe we can also add the Bernoulli/Binomial distribution to said section, so that linking to them makes more sense.

What do you think?

awinml replied:
Yeah, I also feel it should be renamed to "Generalized Linear Models". It will make the docs more consistent and easy to understand. I will also add a section on Bernoulli and Binomial Distribution.

ArturoAmorQ commented on doc/modules/linear_model.rst:
```diff
@@ -848,25 +848,36 @@ Ridge Regression`_, see the example below.
+    distribution and a Logit link. The numerical output of the logistic
+    regression, which is the predicted probability, can be used as a classifier
+    by applying a threshold (by default 0.5) to it. This is how it is
+    implemented in scikit-learn, so it expects a categorical target, making it
```
```suggestion
    implemented in scikit-learn, so it expects a categorical target, making
```

ArturoAmorQ commented on doc/modules/linear_model.rst:
```diff
@@ -848,25 +848,36 @@ Ridge Regression`_, see the example below.
+    distribution and a Logit link. The numerical output of the logistic
+    regression, which is the predicted probability, can be used as a classifier
+    by applying a threshold (by default 0.5) to it. This is how it is
+    implemented in scikit-learn, so it expects a categorical target, making it
+    the Logistic (Regression) Classifier.
```
```suggestion
    implemented in scikit-learn, so it expects a categorical target, making
    the Logistic Regression a classifier.
```

awinml committed:
Update doc/modules/linear_model.rst

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

awinml committed:
Rename  to

awinml committed:
Rename Generalized Linear Regression to Generalized Linear Models

awinml committed:
Added Deviance for Bernoulli and Binomial Distributions

awinml committed:
Added graphs of PMFs of Binomial and Bernoulli Distributions

awinml committed:
Update doc/modules/linear_model.rst

awinml committed:
Merge branch 'main' into doc_logistic_regression

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -1088,9 +1099,9 @@ to warm-starting (see :term:`Glossary <warm_start>`).
         Multinomial Regression." <1311.6529>`
 
-.. _Generalized_linear_regression:
+.. _Generalized_linear_models:
```
Maybe we can try to preserve the old reference identifier in order to not break old URLs with that anchor value for links coming from external websites.

```suggestion
.. _Generalized_linear_regression:

.. _Generalized_linear_models:
```

I am not 100% sure this is authorized in sphinx though. Let's try if it works in the rendered HTML on the CI.

awinml replied:
I tried this. Sphinx does not throw any errors and it works as expected. I will make this change.

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -1088,9 +1099,9 @@ to warm-starting (see :term:`Glossary <warm_start>`).
+.. _Generalized_linear_models:
 
-Generalized Linear Regression
+Generalized Linear Models
 =============================
```
```suggestion
Generalized Linear Models
=========================
```

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -1146,6 +1167,9 @@ The choice of the distribution depends on the problem at hand:
   you might try an Inverse Gaussian deviance (or even higher variance powers
   of the Tweedie family).
+* If the target values :math:`y` are probabilities, you can use the Binomial or
+  Bernoulli deviance. The Binomial/ Bernoulli Deviance with a logit link can be
+  used for binary classification. 
```
```suggestion
  Bernoulli deviance. The Binomial / Bernoulli Deviance with a logit link can be
  used for binary classification. 
```

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -1156,9 +1180,13 @@ Examples of use cases include:
 * Predictive maintenance: number of production interruption events per year
   (Poisson), duration of interruption (Gamma), total interruption time per year
   (Tweedie / Compound Poisson Gamma).
+* Medical Drug Testing: number of cured patients in a set of trials/ number of
+  patients experiencing side effects (Binomial)
```
```suggestion
* Medical Drug Testing: number of cured patients in a set of trials / number of
  patients experiencing side effects (Binomial)
```

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -848,25 +848,36 @@ Ridge Regression`_, see the example below.
+.. note:: **Logistic Regression as a special case of the Generalized Linear Models (GLM)**
+
+    Logistic regression is a special case of
+    :ref:`generalized_linear_models` with a Binomial/Bernoulli conditional
```
```suggestion
    :ref:`generalized_linear_models` with a Binomial / Bernoulli conditional
```

ogrisel reviewed (commented):
Thanks for the PR.

Please find some specific comments below. On the global picture I have the feeling that we should not mention the Binomial distribution because I don't see how you would fit a scikit-learn model on this kind of data for `n > 1`.

On the other hand, we should mention the categorical distribution, that is multinomial with n=1, with the softmax link function that is supported in scikit-learn.

While you are at it, could please fix the broken formatting of the link to the statsmodels package in this page?

awinml committed:
Update doc/modules/linear_model.rst

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

awinml committed:
Update doc/modules/linear_model.rst

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

awinml committed:
Update doc/modules/linear_model.rst

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

awinml committed:
Update doc/modules/linear_model.rst

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

awinml committed:
Add old reference identifier for Generalized_Linear_Regression

awinml committed:
Fix formatting of link to statsmodels package

awinml committed:
Removed mention of Binomial Distribution, Added information for Categorical Distribution

awinml commented:
I have made the following changes:
- Removed all the mentions of Binomial Distribution and have added the required information for Categorical Distribution.
- There were a slight mistake in the formula for Unit Deviance for Gamma distribution, which has been rectified.
- The broken formatting of the link to the statsmodels package has been fixed.

Can you please kindly review the changes? @ogrisel

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -1111,17 +1124,18 @@ The minimization problem becomes:
+================= ================================  ============================================
+Normal            :math:`y \in (-\infty, \infty)`   :math:`(y-\hat{y})^2`
+Bernoulli         :math:`y \in \{0, 1\}`            :math:`2({y}\log\frac{y}{\hat{y}}+({1}-{y})\log\frac{{1}-{y}}{{1}-\hat{y}})`
+Categorical       :math:`y \in \{0, 1, ..., k-1\}`  :math:`2\sum_{i=0}^{k_\text{classes}-1} y_\text{i}\log\frac{y_\text{i}}{\hat{y_\text{i}}}`
```
In this notation, `y_i` is an (ordinal class label) integer, so it does not work. Think you need to use a binary class indicator variable.

ArturoAmorQ commented on doc/modules/linear_model.rst:
```diff
@@ -848,25 +848,36 @@ Ridge Regression`_, see the example below.
-classifier. In this model, the probabilities describing the possible outcomes
-of a single trial are modeled using a
-`logistic function <https://en.wikipedia.org/wiki/Logistic_function>`_.
+LogisticRegression, despite its name, is implemented as a linear model for
+classification rather than regression in terms of scikit-learn/ML nomenclature.
+Logistic regression is also known in the literature as logit regression,
+maximum-entropy classification (MaxEnt) or the log-linear classifier. In this
+model, the probabilities describing the possible outcomes of a single trial are
+modeled using a `logistic function
+<https://en.wikipedia.org/wiki/Logistic_function>`_.
 
 Logistic regression is implemented in :class:`LogisticRegression`.
 This implementation can fit binary, One-vs-Rest, or multinomial logistic
 regression with optional :math:`\ell_1`, :math:`\ell_2` or Elastic-Net
 regularization.
```
```suggestion
The logistic regression is implemented in :class:`LogisticRegression`. Despite
its name, it is implemented as a linear model for classification rather than
regression in terms of the scikit-learn/ML nomenclature. The logistic
regression is also known in the literature as logit regression,
maximum-entropy classification (MaxEnt) or the log-linear classifier. In this
model, the probabilities describing the possible outcomes of a single trial
are modeled using a `logistic function
<https://en.wikipedia.org/wiki/Logistic_function>`_.

This implementation can fit binary, One-vs-Rest, or multinomial logistic
regression with optional :math:`\ell_1`, :math:`\ell_2` or Elastic-Net
regularization.
```

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -1156,9 +1182,13 @@ Examples of use cases include:
   (Poisson), duration of interruption (Gamma), total interruption time per year
   (Tweedie / Compound Poisson Gamma).
+* Medical Drug Testing: number of cured patients in a set of trials / number of
+  patients experiencing side effects (Bernoulli)
```
Same comment here. Maybe you could also introduce a 3 class classification problem with mutually exclusive classes, e.g. sentiment analysis with positive, neutral and negative sentiments. Although, in this case, one could argue that this is actually original regression because of the ordering of the 3 classes. Maybe there is another simple example where the classes are not naturally ordered.

awinml replied:
One example that I could think of is that of a News Classification Problem where we are classifying news articles into 3 categories namely Business News, Politics and Entertainment news.

ogrisel reviewed (approved):
A few more remark but otherwise, LGTM. Maybe @lorentzenchr might be interested in giving this PR a review.

awinml committed:
Update doc/modules/linear_model.rst

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

lorentzenchr commented on doc/modules/linear_model.rst:
```diff
@@ -1135,17 +1149,29 @@ in the following figure,
    distribution, but not for the Gamma distribution which has a strictly
    positive target domain.
+   
+.. figure:: ./glm_data/bernoulli_distribution.png
```
This figure should be improved. If `Y` is categorical, the x-axis should reflect that.
I'm not sure, if we need this figure at all.

awinml replied:
I added this figure for completeness. But I do see your point, this figure is not needed. A short note about the Bernoulli distribution instead would be a better idea.

lorentzenchr commented on doc/modules/linear_model.rst:
```diff
@@ -1135,17 +1149,29 @@ in the following figure,
+   :scale: 100%
+
+   PMF of a random variable Y following Bernoulli Distributions with probability
+   of success (p). Bernoulli Distribution is a special case of the Categorical
```
```suggestion
   of success (p=0.25). Bernoulli Distribution is a special case of the Categorical
```

awinml committed:
Update doc/modules/linear_model.rst

awinml committed:
Update doc/modules/linear_model.rst

awinml committed:
Update doc/modules/linear_model.rst

ArturoAmorQ commented:
Hi @awinml. Now that you removed the reference to the figure, you still have to remove the file [doc/modules/glm_data/bernoulli_distribution.png](https://github.com/scikit-learn/scikit-learn/pull/24738/files#diff-977c026c455afd560e1f14e2a159d62c660f607f3cb958e549de66d6ef5d3106).

Also please notice that each bullet point in the examples of use cases end with a point.

awinml committed:
Update doc/modules/linear_model.rst

awinml committed:
Merge branch 'main' into doc_logistic_regression

awinml commented:
@ArturoAmorQ  I have removed the file [doc/modules/glm_data/bernoulli_distribution.png](https://github.com/scikit-learn/scikit-learn/pull/24738/files#diff-977c026c455afd560e1f14e2a159d62c660f607f3cb958e549de66d6ef5d3106). That was an oversight from my part. I also added the missing full-stops for all the examples of use cases that have been added.

lorentzenchr commented:
@ArturoAmorQ Do you want to approve this PR? I will merge with your +1.

ArturoAmorQ commented on doc/modules/linear_model.rst:
```diff
@@ -1156,10 +1182,15 @@ Examples of use cases include:
+* Medical Drug Testing: probability of curing a patient in a set of trials /
+  probability that a patient will experience side effects (Bernoulli).
+* News Classification: Classification of news articles into three categories
+  namely Business News, Politics and Entertainment news. (Categorical).
```
```suggestion
  namely Business News, Politics and Entertainment news (Categorical).
```

ArturoAmorQ reviewed (approved):
Just a nitpick, otherwise LGTM. Thanks for your time and patience @awinml!

awinml committed:
Update doc/modules/linear_model.rst

Co-authored-by: Arturo Amor <86408019+ArturoAmorQ@users.noreply.github.com>

lorentzenchr commented on doc/modules/linear_model.rst:
```diff
@@ -1156,10 +1182,15 @@ Examples of use cases include:
   (Poisson), duration of interruption (Gamma), total interruption time per year
   (Tweedie / Compound Poisson Gamma).
-
+* Medical Drug Testing: probability of curing a patient in a set of trials /
```
```suggestion
* Medical Drug Testing: probability of curing a patient in a set of trials or
```

lorentzenchr commented on doc/modules/linear_model.rst:
```diff
@@ -1156,10 +1182,15 @@ Examples of use cases include:
 * Risk modeling / insurance policy pricing:  number of claim events /
   policyholder per year (Poisson), cost per event (Gamma), total cost per
   policyholder per year (Tweedie / Compound Poisson Gamma).
+* Credit Default / Fraud Detection: number of fraudulent transactions from a
+  region (Bernoulli).
```
@ogrisel already raised concerns about "number". Should it be "probability"?

awinml committed:
Update doc/modules/linear_model.rst

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

awinml committed:
Update doc/modules/linear_model.rst

awinml committed:
Update doc/modules/linear_model.rst
